### DISCUSSION TITLE
bit.band in v6.0.26

### DISCUSSION THREAD

**Discussion by Godlike-Productions** [RomanKubiak/ctrlr#206] [Q&A] [answered]
Does anyone know the code for bitwise AND function in 6.0.26?

bit.band doesn't seem to work anymore.

Alternatively does anyone know the library that was used in lua 5.1 for Ctrlr in earlier version, so that I can reimport?

![bitband](https://user-images.githubusercontent.com/13851825/110226454-68e31a80-7f2a-11eb-9f46-3405fa65cc9b.png)

**Comment by dnaldoog** [comment]
I think the error message would say **a nil value**, not- **function value** if it doesn’t exist. What is the code snippet you are using?

**Reply by Godlike-Productions** [reply]
Here's a little bit of the code.  I use them a lot for switches that turn individual bits on and off in the sysex byte.

```
-- Read LFO Waveform Slider
			lfoWaveformSliderValue = panel:getModulatorByName("lfoWaveformSlider"):getValue()
			if bit.band(midiBuffer[58],3) ~= lfoWaveformSliderValue then
 			   	byte58 = bit.band(midiBuffer[58],252)+lfoWaveformSliderValue
  		 		midiBuffer[58] = byte58
				bufferChanged = true
			end

-- Read VCO Mod Switch
			vcoModSwitchValue = panel:getModulatorByName("vcoModSwitch"):getValueMapped() --ValueMapped may not work with image sliders - check
			if bit.band(midiBuffer[58],12) ~= vcoModSwitchValue then
 			   	byte58 = bit.band(midiBuffer[58],243)+vcoModSwitchValue
  		 		midiBuffer[58] = byte58
				bufferChanged = true
			end
			
```

This used to work, and I'm pretty sure it even worked in this version of ctrlr, but as soon as I enable the method to poll my controllers it crashes.

This is the start of the timer callback

```
function timerCallback(timerId)

-- this function formats midiBuffer as a Edit Buffer Load. Patch data starts at byte 7. Don't use for 
-- patch dumps to specific patch locations

-- byte's 07 and 08 are the patch protected byte. 0xff is protected, 0x00 is not-protected. Use 0x00

	-- Load program timer
	if timerId == 1 then
		if lastMidiBuffer == {0x05} then -- this can be removed. It's handled differently now.
		--	timer:stopTimer(timerId)
			return
		else
		-- Read VCO2 Range Slider
			vco2RSliderValue = panel:getModulatorByName("vco2RangeSlider"):getValue()
			vco2Rmsb = math.floor(vco2RSliderValue/16)
			vco2Rlsb = vco2RSliderValue%16
			if midiBuffer[9] ~= vco2Rmsb then
				bufferChanged = true
				midiBuffer[9] = vco2Rmsb
			end
			if midiBuffer[10] ~= vco2Rlsb then
				bufferChanged = true
				midiBuffer[10] = vco2Rlsb
			end
		
```

**Comment by Godlike-Productions** [comment, marked as answer]
I've fixed it.

I've saved this module - https://github.com/davidm/lua-bit-numberlua/blob/master/lmod/bit/numberlua.lua as a LUA file and removed the word local from before M on line 183.

M.bit.band now works.  I'll go through and update my code to use this.

I could probably also bring it in as a module, but I'm not sure if modules export into restricted instances.

**Comment by dnaldoog** [comment]
I think bit.band does work in version 6.02!

[bit.band exists in 6.02_1_0_Hell-O-Kitty_2021-03-07_18-05.panel.zip](https://github.com/RomanKubiak/ctrlr/files/6096853/bit.band.exists.in.6.02_1_0_Hell-O-Kitty_2021-03-07_18-05.panel.zip)

![bb6026](https://user-images.githubusercontent.com/76831683/110234855-36f4a700-7f70-11eb-8d6a-8aa11e2e6213.png)

**Reply by Godlike-Productions** [reply]
I have no idea why it failed in my panel then.  I have crashed it pretty hard a few times.  Maybe something got shaken loose.

**Comment by dobo365** [comment]
There are differences in bit operations depending on the Lua version.

The following bitwise operations are possible in Lua 5.1:
─	bit.tohex (x[,n])	-- convert x to hex with n digits (default 8)
─	bit.bnot (x)		-- bitwise not of x
─	bit.band (x1[,x2...])	-- bitwise and of x1, x2, ...
─	bit.bor (x1[,x2...])	-- bitwise or of x1, x2, ...
─	bit.bxor (x1[,x2...])	-- bitwise xor of x1, x2, ...
─	bit.lshift (x, n)	-- left-shift of x by n bits
─	bit.rshift (x, n)	-- logical right-shift of x by n bits
─	bit.arshift (x, n)	-- arithmetic right-shift of x by n bits
─	bit.rol (x, n)		-- left-rotate of x by n bits
─	bit.ror (x, n)		-- right-rotate of x by n bits
─	bit.bswap (x)	-- byte-swap of x (little-endian <-> big-endian)

With Lua 5.3, it will be possible to use the built-in operators:
─	&: bitwise and
─	|: bitwise or
─	~: bitwise exclusive or
─	>>: right shift
─	<<: left shift
─	~: unary bitwise not

In Ctrlr 6.0.26, it is maybe Lua 5.3...
In Ctrlr 5.3.201 that I'm using it is Lua 5.1

**Reply by Godlike-Productions** [reply]
Yeah, it wasn't working for some reason, but the other example did in the same version.

I'll check my code to see if bit got redefined somewhere.

**Reply by Possemo** [reply]
> In Ctrlr 6.0.26, it is maybe Lua 5.3...
> In Ctrlr 5.3.201 that I'm using it is Lua 5.1

No - look at the Ctrlr/About window. Lua 5.2 makes a big step forward. Roman once tried to make the jump but he had to revert back to 5.1 - v5.2 is too different. We probably won't see anything else than Lua 5.1 for a long time. For me it's ok. I don't miss anything.

**Reply by Godlike-Productions** [reply]
Yep 6.0.26 uses Lua 5.1.4.  The binary functions don't have shorthand like >> or << or ||, but it's really not a hassle.

**Comment by Godlike-Productions** [comment]
I think I found an issue.  I was experimenting with base64 encoding (to encode memoryBlocks as strings to support json export), and used a base64Methods library.  They have a function called bit.

**Comment by mtarenskeen** [comment]
In some of my panels I use my own LUA function to replace bit.and and the likes, because I wanted it to work on many different Ctrlr and LUA versions, including older ones.

```
function bitband(a, b)
    local result = 0
    local bitval = 1
    while a > 0 and b > 0 do
      if a % 2 == 1 and b % 2 == 1 then -- test the rightmost bits
          result = result + bitval      -- set the current bit
      end
      bitval = bitval * 2 -- shift left
      a = math.floor(a/2) -- shift right
      b = math.floor(b/2)
    end
    return result
end


```
